### ISSUE TITLE
mistral example seems to hang on Loading model from disk.

### ISSUE THREAD

**Issue by Diniden** [ml-explore/mlx-examples#25]
I am trying to run the mistral model 7B following the README's instructions (and adding the missing numpy dep)

This is all I see for a very long time

```
> python mistral.py
[INFO] Loading model from disk.
```

Hardware: Apple M1 Max
OS: 14.1.2 (23B92)
Python 3.11.5
conda 23.7.4

My python process stays at a really low CPU and I am not seeing any disk access that is worthy of fiddling with 14GB of data.

Any way to better debug this or get to something working?

**Comment by Diniden** [comment]
Printed out logs and found it hangs here as probably expected. Unsure how to debug why the mlx library is hanging with no work getting done here:

```
weights = mx.load(str(model_path / "mlx_mistral_7b.npz"))
```

I am reading into the mlx library now to see if there are further steps I can take

**Comment by awni** [comment]
Strange.. I just ran it no problem on an M1 Max. Is it possible the weights are ill formatted? Maybe the conversion didn't work?

**Comment by awni** [comment]
If this is still an issue, please reopen.

**Comment by Diniden** [comment]
Definitely still an issue. I just ran the script through the night and it stayed at the same hang point.

Not sure how I can re-open this issue. Unless you mean make another post.

I'm going to try a debugger and step through it, or compile the mlx lib and see if I can find out what's wrong.

I followed the README's instructions exactly and repeated the weight conversion step multiple times now with no luck.

I will also see if I can make sure the mistral download has matching hash in case the download flubbed...

**Comment by Blucknote** [comment]
I'm experience kinda same, after following README it hangs with no output for more than 10 minutes but for me it has additional row:
```
python3 mistral.py --prompt "It is a truth universally acknowledged,"  --temp 0
[INFO] Loading model from disk.
[INFO] Starting generation...
It is a truth universally acknowledged,
```

After that no significal CPU/GPU usage and i have to kill process.

Enviroinment info:
M1 Pro 16Gb
Sonoma 14.1.1 (23B81)
Python 3.9.5

Also strange that pip cannot find mlx package for python3.10 3.11 and 3.12 despite PyPi requires python 3.7+

**Comment by awni** [comment]
@Diniden we've updated MLX and fixed a bug with loading from NumPy so try `pip install --upgrade mlx` and see if that helps. I would also double check the weights downloaded properly.

@Blucknote it looks like you are stuck in a very different spot. It would be helpful if you could fine more precisely where it is stuck by stepping through with PDB or adding some prints around [this part of the code](https://github.com/ml-explore/mlx-examples/blob/main/mistral/mistral.py#L263-L270).

Also regarding installation, you should definitely be able to install MLX with pip with a higher version of Python. The [install docs](https://ml-explore.github.io/mlx/build/html/install.html) have more information on possible issues.

**Comment by Diniden** [comment]
New version worked (sort of)!

The new release properly gave me:

libc++abi: terminating due to uncaught exception of type pybind11::error_already_set: BadZipFile: Bad CRC-32 for file 'layers.24.feed_forward.w3.weight.npy'

Instead of silent failure and infinite hang, which was really nice.

I re-downloaded the mistral weights and performed the steps again, and THEN it worked flawlessly!

Thanks for looking at this.

**Comment by Blucknote** [comment]
Patience I must have: in debugger I have output,  the most long time row is `y = sample(logits[:, -1, :])`
may be on quantized weights it will be faster but this check for smarter person

**Comment by awni** [comment]
Hmm, I wonder why it's so slow. I assume you are using the Metal back-end? 

```
python -c "import mlx.core as mx; print(mx.default_device())"
```

Should give `Device(gpu, 0)`.

If it's using the GPU, maybe there is some swapping, I am testing on a 32GB machine and it processes the prompt very quickly < second.

**Comment by Blucknote** [comment]
> Should give `Device(gpu, 0)`.

Output as expected: 
```python
import mlx.core as mx;
PyDev console: starting.
print(mx.default_device())
Device(gpu, 0)
```
swap is defenitely using

However... can i trust htop on mac? :D System monitor gives me 14Gb Ram usage and 12Gb swap
but htop says it's 1.3 Gb mem and ~12 Gb on swap

[Stats](https://github.com/exelban/stats) shows same as system monitor

**Comment by eugenepyvovarov** [comment]
> I'm experience kinda same, after following README it hangs with no output for more than 10 minutes but for me it has additional row:
> 
> ```
> python3 mistral.py --prompt "It is a truth universally acknowledged,"  --temp 0
> [INFO] Loading model from disk.
> [INFO] Starting generation...
> It is a truth universally acknowledged,
> ```
> 
> After that no significal CPU/GPU usage and i have to kill process.
> 
> Enviroinment info: M1 Pro 16Gb Sonoma 14.1.1 (23B81) Python 3.9.5
> 
> Also strange that pip cannot find mlx package for python3.10 3.11 and 3.12 despite PyPi requires python 3.7+

experiencing same issue M1 Pro 16gb, stuck after first phrase

**Comment by awni** [comment]
I have a 32GB machine and the Mistral example runs pretty quickly. It seems like somewhere between 16 and 32 is the cutoff right now to get good perf, but I'm not sure exactly where. We're investigating memory use to see if it can be improved.

**Comment by sproblvem** [comment]
> I have a 32GB machine and the Mistral example runs pretty quickly. It seems like somewhere between 16 and 32 is the cutoff right now to get good perf, but I'm not sure exactly where. We're investigating memory use to see if it can be improved.

Thank you guys for this great library. 
But I have also encountered same issue with [[singhaki](https://github.com/singhaki)，](https://github.com/ml-explore/mlx-examples/issues/25#issuecomment-1846550551). The mistral example runs correctly, within 2 hours... My computer is a M2 Pro 16gb two. Here is the memory usage.
![image](https://github.com/ml-explore/mlx-examples/assets/1687281/d93c27c8-bc75-4112-9abf-02d2aca9b63f)

**Comment by bigsnarfdude** [comment]
can confirm m2pro 32gb works without issues. loads and outputs

**Comment by rodchile** [comment]
I observe the same pattern than the one reported by @sproblvem [here](https://github.com/ml-explore/mlx-examples/issues/25#issuecomment-1850111443). When running either Mistral or Llama with 16GB RAM, it runs pretty slow (using either my own build weights or downloaded from the MLX community on Hugging Face). My setup has the configuration [here](https://ml-explore.github.io/mlx/build/html/install.html).

When running the quantized model on either option, performance improved a lot.

**Comment by Blucknote** [comment]
> quantized model

@rodchile can you share more details please? how did you get quantizing? If it was gguf from TheBloke or something other? If it was gguf does conversion script able to convert  to required format?

**Comment by awni** [comment]
@Blucknote There are pre-converted quantized models in the MLX Hugging Face community: https://huggingface.co/mlx-community

Also, all of the conversion scripts in the [LLM examples](https://github.com/ml-explore/mlx-examples/tree/main/llms) can produce quantized models

**Comment by Blucknote** [comment]
@awni thanks!
